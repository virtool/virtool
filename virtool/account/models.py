from datetime import datetime

from email_validator import EmailSyntaxError, validate_email
from pydantic import ConstrainedStr, validator

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.enums import AnalysisWorkflow
from virtool_core.models.group import GroupMinimal, Permissions
from virtool_core.models.user import User


def check_email(email: str | None) -> str | None:
    """Checks if the given email is valid."""
    if email is None:
        return None

    try:
        validate_email(email)
    except EmailSyntaxError:
        raise ValueError("The format of the email is invalid")

    return email


class AccountSettings(BaseModel):
    quick_analyze_workflow: AnalysisWorkflow
    show_ids: bool
    show_versions: bool
    skip_quick_analyze_dialog: bool


class ConstrainedEmail(ConstrainedStr):
    strip_whitespace = True


class Account(User):
    settings: AccountSettings
    email: ConstrainedEmail | None

    _email_validation = validator("email", allow_reuse=True)(check_email)


class APIKey(BaseModel):
    id: str
    created_at: datetime
    groups: list[GroupMinimal]
    name: str
    permissions: Permissions
