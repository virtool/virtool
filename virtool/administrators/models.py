from typing import List

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.roles import AdministratorRole
from virtool_core.models.user import UserNested


class AdministratorMinimal(UserNested):
    role: AdministratorRole


class Administrator(AdministratorMinimal):
    available_roles: List[dict]


class AdministratorSearch(BaseModel):
    items: List[AdministratorMinimal]
    available_roles: List[dict]
