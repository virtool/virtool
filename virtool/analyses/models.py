import enum
from sqlalchemy import Column, Integer, String, DateTime, Enum

from virtool.postgres import Base


class AnalysisFormat(str, enum.Enum):
    """
    Enumerated type for analysis file formats

    """

    sam = "sam"
    bam = "bam"
    fasta = "fasta"
    fastq = "fastq"
    csv = "csv"
    tsv = "tsv"
    json = "json"


class AnalysisFile(Base):
    """
    SQL model to store new analysis files

    """
    __tablename__ = "analysis_files"

    id = Column(Integer, primary_key=True)
    analysis = Column(String)
    description = Column(String)
    format = Column(Enum(AnalysisFormat))
    name = Column(String)
    name_on_disk = Column(String, unique=True)
    size = Column(Integer)
    uploaded_at = Column(DateTime)

    def __repr__(self):
        return f"<AnalysisFile(id={self.id}, analysis={self.analysis}, description={self.description}, " \
               f"format={self.format}, name={self.name}, name_on_disk={self.name_on_disk}, size={self.size}, " \
               f"uploaded_at={self.uploaded_at})>"
