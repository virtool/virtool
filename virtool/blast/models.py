from datetime import datetime

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.task import TaskNested


class NuvsBlast(BaseModel):
    id: int
    created_at: datetime
    updated_at: datetime
    last_checked_at: datetime
    error: str | None
    rid: str | None
    ready: bool = False
    result: dict | None
    task: TaskNested | None
