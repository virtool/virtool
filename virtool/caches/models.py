from sqlalchemy import Column, DateTime, Enum, Integer, String
from sqlalchemy.sql.schema import UniqueConstraint

from virtool.pg.utils import Base
from virtool.samples.models import ArtifactType


class SampleArtifactCache(Base):
    """
    SQL model to store a cached sample artifact

    """
    __tablename__ = "sample_artifacts_cache"
    __table_args__ = (UniqueConstraint("key", "name", "sample"),)

    id = Column(Integer, primary_key=True)
    key = Column(String, nullable=False)
    name = Column(String, nullable=False)
    name_on_disk = Column(String)
    sample = Column(String, nullable=False)
    size = Column(Integer)
    type = Column(Enum(ArtifactType), nullable=False)
    uploaded_at = Column(DateTime)


class SampleReadsCache(Base):
    """
    SQL model to store cached sample reads files

    """
    __tablename__ = "sample_reads_cache"
    __tableargs__ = (UniqueConstraint("key", "name", "sample"),)

    id = Column(Integer, primary_key=True)
    key = Column(String, nullable=False)
    name = Column(String(length=13), nullable=False)
    name_on_disk = Column(String, nullable=False)
    sample = Column(String, nullable=False)
    size = Column(Integer)
    uploaded_at = Column(DateTime)