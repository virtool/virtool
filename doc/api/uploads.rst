.. _uploads:

Uploads
=======

When uploads are required in Virtool, they should be targeted at ``/upload``. The available endpoints are:

+-------------------------------------------------+-------------------------+
| Situation                                       | Endpoint                |
+=================================================+=========================+
| Upload and import a Virtool viruses export file | /upload/viruses         |
+-------------------------------------------------+-------------------------+
| Upload a Illumina read file                     | /upload/reads           |
+-------------------------------------------------+-------------------------+
| Upload a .hmm file for use with NuVs            | /upload/hmm/profiles    |
+-------------------------------------------------+-------------------------+
| Upload and import a NuVs annotation file        | /upload/hmm/annotations |
+-------------------------------------------------+-------------------------+
| Upload a host FASTA file                        | /upload/host            |
+-------------------------------------------------+-------------------------+

