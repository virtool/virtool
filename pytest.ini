[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
