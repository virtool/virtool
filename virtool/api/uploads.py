import os

from cerberus import Validator

import virtool.db.files
import virtool.http.routes
import virtool.utils
from virtool.api.utils import invalid_query, json_response, not_found

CHUNK_SIZE = 131072

FILE_TYPES = [
    "reference",
    "reads",
    "hmm",
    "subtraction"
]

routes = virtool.http.routes.Routes()


@routes.post("/upload/{file_type}", permission="upload_file")
async def upload(req):
    file_type = req.match_info["file_type"]

    if file_type not in FILE_TYPES:
        return not_found()

    v = Validator({
        "name": {"type": "string", "required": True}
    }, allow_unknown=True)

    if not v.validate(dict(req.query)):
        return invalid_query(v.errors)

    reader = await req.multipart()
    file = await reader.next()

    db = req.app["db"]

    filename = req.query["name"]

    document = await virtool.db.files.create(
        db,
        filename,
        file_type,
        req["client"].user_id
    )

    file_path = os.path.join(req.app["settings"].get("data_path"), "files", document["id"])

    size = 0

    with open(file_path, "wb") as handle:
        while True:
            chunk = await file.read_chunk(CHUNK_SIZE)
            if not chunk:
                break
            size += len(chunk)
            handle.write(chunk)

    return json_response(document, status=201)
