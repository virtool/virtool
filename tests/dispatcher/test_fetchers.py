import pytest

from virtool.dispatcher.change import Change
from virtool.dispatcher.fetchers import IndexesFetcher, SimpleMongoFetcher, LabelsFetcher, UploadsFetcher, TasksFetcher
from virtool.dispatcher.operations import DELETE, INSERT, UPDATE
from virtool.uploads.models import UploadType


@pytest.fixture
def connections(ws):
    return [ws, ws, ws]


class TestSimpleMongoFetcher:

    async def test_auto_delete(self, connections, dbi, ws):
        fetcher = SimpleMongoFetcher(dbi.hmm)

        pairs = list()

        async for pair in fetcher.fetch(Change("hmm", DELETE, ["foo"]), connections):
            pairs.append(pair)

        message = {
            "interface": "hmm",
            "operation": DELETE,
            "data": ["foo"]
        }

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]

    @pytest.mark.parametrize("operation", [INSERT, UPDATE])
    @pytest.mark.parametrize("use_projection", [True, False])
    async def test_insert_and_update(self, operation, use_projection, connections, dbi, ws):
        await dbi.hmm.insert_many([
            {"_id": "foo", "name": "Foo", "include": True, "hidden": True},
            {"_id": "bar", "name": "Bar", "include": True, "hidden": True}
        ])

        if use_projection:
            fetcher = SimpleMongoFetcher(dbi.hmm, projection=["name", "include"])
        else:
            fetcher = SimpleMongoFetcher(dbi.hmm)

        pairs = list()

        async for pair in fetcher.fetch(Change("hmm", operation, ["foo"]), connections):
            pairs.append(pair)

        message = {
            "interface": "hmm",
            "operation": operation,
            "data": {
                "id": "foo",
                "name": "Foo",
                "include": True
            }
        }

        # If not using projection, the `hidden` field will be included in the message.
        if not use_projection:
            message["data"]["hidden"] = True

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]


class TestIndexesFetcher:

    async def test_auto_delete(self, connections, dbi, ws):
        fetcher = IndexesFetcher(dbi)

        pairs = list()

        async for pair in fetcher.fetch(Change("indexes", DELETE, ["foo.3"]), connections):
            pairs.append(pair)

        message = {
            "interface": "indexes",
            "operation": DELETE,
            "data": ["foo.3"]
        }

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]

    @pytest.mark.parametrize("operation", [INSERT, UPDATE])
    async def test_insert_and_update(
            self,
            operation,
            mocker,
            connections,
            dbi,
            reference,
            static_time,
            test_indexes,
            ws
    ):
        await dbi.indexes.insert_many(test_indexes)

        async def m_indexes_processor(db, document):
            assert db == dbi

            processed = dict(document)
            processed["id"] = processed.pop("_id")
            processed["extra_field"] = True

            return processed

        mocker.patch("virtool.indexes.db.processor", m_indexes_processor)

        fetcher = IndexesFetcher(dbi)

        pairs = list()

        async for pair in fetcher.fetch(Change("indexes", operation, ["cdffbdjk"]), connections):
            pairs.append(pair)

        message = {
            "data": {
                "created_at": test_indexes[1]["created_at"],
                "extra_field": True,
                "has_files": False,
                "id": "cdffbdjk",
                "job": {
                    "id": "3tt0w336"
                },
                "ready": True,
                "reference": {
                    "id": "hxn167"
                },
                "user": {
                    "id": "mrott"
                },
                "version": 1
            },
            "interface": "indexes",
            "operation": operation
        }

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]


class TestLabelsFetcher:

    async def test_auto_delete(self, connections, dbi, pg, ws):
        fetcher = LabelsFetcher(pg, dbi)

        pairs = list()

        message = {
            "interface": "labels",
            "operation": DELETE,
            "data": [1]
        }

        async for pair in fetcher.fetch(Change("labels", DELETE, [1]), connections):
            pairs.append(pair)

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]

    @pytest.mark.parametrize("operation", [INSERT, UPDATE])
    async def test_insert_and_update(
            self,
            operation,
            connections,
            dbi,
            pg,
            reference,
            static_time,
            test_labels,
            ws
    ):
        fetcher = LabelsFetcher(pg, dbi)

        pairs = list()

        async for pair in fetcher.fetch(Change("labels", operation, [1]), connections):
            pairs.append(pair)

        message = {
            "data": {
                "id": 1,
                "name": "Legacy",
                "color": None,
                "description": None
            },
            "interface": "labels",
            "operation": operation
        }

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]


class TestUploadsFetcher:

    async def test_auto_delete(self, connections, pg, ws):
        fetcher = UploadsFetcher(pg)

        pairs = list()

        message = {
            "interface": "uploads",
            "operation": DELETE,
            "data": [1]
        }

        async for pair in fetcher.fetch(Change("uploads", DELETE, [1]), connections):
            pairs.append(pair)

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]

    @pytest.mark.parametrize("operation", [INSERT, UPDATE])
    async def test_insert_and_update(
            self,
            operation,
            connections,
            pg,
            reference,
            static_time,
            test_uploads,
            ws
    ):
        fetcher = UploadsFetcher(pg)

        pairs = list()

        async for pair in fetcher.fetch(Change("uploads", operation, [1]), connections):
            pairs.append(pair)

        message = {
            "data": {
                "id": 1,
                "created_at": None,
                "name": "test.fq.gz",
                "name_on_disk": None,
                "ready": False,
                "removed": False,
                "removed_at": None,
                'reserved': False,
                "size": None,
                "type": UploadType.reads,
                "user": "danny",
                "uploaded_at": None
            },
            "interface": "uploads",
            "operation": operation
        }

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]


class TestTasksFetcher:

    async def test_auto_delete(self, connections, pg, ws):
        fetcher = TasksFetcher(pg)

        pairs = list()

        message = {
            "interface": "tasks",
            "operation": DELETE,
            "data": [1]
        }

        async for pair in fetcher.fetch(Change("tasks", DELETE, [1]), connections):
            pairs.append(pair)

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]

    @pytest.mark.parametrize("operation", [INSERT, UPDATE])
    async def test_insert_and_update(
            self,
            operation,
            connections,
            pg,
            reference,
            static_time,
            test_tasks,
            ws
    ):
        fetcher = TasksFetcher(pg)

        pairs = list()

        async for pair in fetcher.fetch(Change("tasks", operation, [1]), connections):
            pairs.append(pair)

        message = {
            "interface": "tasks",
            "operation": operation,
            "data": {
                "id": 1,
                "complete": True,
                "context": None,
                "count": 0,
                "created_at": static_time.datetime,
                "error": None,
                "file_size": None,
                "progress": 100,
                "step": "download",
                "type": "clone_reference"
            }
        }

        assert pairs == [
            (ws, message),
            (ws, message),
            (ws, message)
        ]
