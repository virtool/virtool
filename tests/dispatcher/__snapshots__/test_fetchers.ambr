# name: TestUploadsFetcher.test_insert_and_update[uvloop-insert]
  <class 'list'> [
    <class 'dict'> {
      'data': <class 'dict'> {
        'created_at': None,
        'id': 1,
        'name': 'test.fq.gz',
        'name_on_disk': None,
        'ready': False,
        'removed': False,
        'removed_at': None,
        'reserved': False,
        'size': None,
        'type': <UploadType.reads: 'reads'>,
        'uploaded_at': None,
        'user': <class 'dict'> {
          'administrator': False,
          'handle': 'leeashley',
          'id': '7CtBo2yG',
        },
      },
      'interface': 'uploads',
      'operation': 'insert',
    },
    <class 'dict'> {
      'data': <class 'dict'> {
        'created_at': None,
        'id': 1,
        'name': 'test.fq.gz',
        'name_on_disk': None,
        'ready': False,
        'removed': False,
        'removed_at': None,
        'reserved': False,
        'size': None,
        'type': <UploadType.reads: 'reads'>,
        'uploaded_at': None,
        'user': <class 'dict'> {
          'administrator': False,
          'handle': 'leeashley',
          'id': '7CtBo2yG',
        },
      },
      'interface': 'uploads',
      'operation': 'insert',
    },
    <class 'dict'> {
      'data': <class 'dict'> {
        'created_at': None,
        'id': 1,
        'name': 'test.fq.gz',
        'name_on_disk': None,
        'ready': False,
        'removed': False,
        'removed_at': None,
        'reserved': False,
        'size': None,
        'type': <UploadType.reads: 'reads'>,
        'uploaded_at': None,
        'user': <class 'dict'> {
          'administrator': False,
          'handle': 'leeashley',
          'id': '7CtBo2yG',
        },
      },
      'interface': 'uploads',
      'operation': 'insert',
    },
  ]
---
# name: TestUploadsFetcher.test_insert_and_update[uvloop-update]
  <class 'list'> [
    <class 'dict'> {
      'data': <class 'dict'> {
        'created_at': None,
        'id': 1,
        'name': 'test.fq.gz',
        'name_on_disk': None,
        'ready': False,
        'removed': False,
        'removed_at': None,
        'reserved': False,
        'size': None,
        'type': <UploadType.reads: 'reads'>,
        'uploaded_at': None,
        'user': <class 'dict'> {
          'administrator': False,
          'handle': 'leeashley',
          'id': '7CtBo2yG',
        },
      },
      'interface': 'uploads',
      'operation': 'update',
    },
    <class 'dict'> {
      'data': <class 'dict'> {
        'created_at': None,
        'id': 1,
        'name': 'test.fq.gz',
        'name_on_disk': None,
        'ready': False,
        'removed': False,
        'removed_at': None,
        'reserved': False,
        'size': None,
        'type': <UploadType.reads: 'reads'>,
        'uploaded_at': None,
        'user': <class 'dict'> {
          'administrator': False,
          'handle': 'leeashley',
          'id': '7CtBo2yG',
        },
      },
      'interface': 'uploads',
      'operation': 'update',
    },
    <class 'dict'> {
      'data': <class 'dict'> {
        'created_at': None,
        'id': 1,
        'name': 'test.fq.gz',
        'name_on_disk': None,
        'ready': False,
        'removed': False,
        'removed_at': None,
        'reserved': False,
        'size': None,
        'type': <UploadType.reads: 'reads'>,
        'uploaded_at': None,
        'user': <class 'dict'> {
          'administrator': False,
          'handle': 'leeashley',
          'id': '7CtBo2yG',
        },
      },
      'interface': 'uploads',
      'operation': 'update',
    },
  ]
---
